grade_test.go

```diff
@@ -362,33 +389,35 @@ func TestGradeSubmission(t *testing.T) {
 	os.Create(dir + "/main.cpp")
 	p := path.Join(workDir, dir, "main.cpp")
 	os.WriteFile(p, []byte(`#include <iostream>
-	int main() { std::cout << "Hello world!" << std::endl; }`), 0666)
+	int main() { std::cout << "Hello World!"; return 0;}`), 0666)
 
 	if err != nil {
 		t.Fatalf("Failed to make temp dir %#v", err)
 	}
 
 	// run and validate
 	runArgs := ""
-	expected := "Hello world!"
+	expected := "Hello World!"
+	language := "c++"
 	input := []string{}
 	wall := false
+	result := util.SubmissionResult{Student: "jgg0144", CompileSuccess: false, RunCorrect: false, Feedback: "", AssignmentId: 1, StudentId: 1}
 
-	actual := gradeSubmission(dir, workDir, runArgs, expected, input, wall)
+	gradeSubmission(&result, dir, workDir, runArgs, expected, language, input, wall)
 
-	if !actual.CompileSuccess {
+	if !result.CompileSuccess {
 		t.Fatalf("Compile error")
 	}
 
-	if actual.Feedback != " Hello world!" {
-		t.Errorf("actual.diff mismatch, received %#v, want %#v", actual.Feedback, " Hello world!")
+	if result.Feedback != " Hello World!" {
+		t.Errorf("actual.diff mismatch, received %#v, want %#v", result.Feedback, " Hello World!")
 	}
 
-	if !actual.RunCorrect {
+	if !result.RunCorrect {
 		t.Errorf("actual.runCorrect is false, want true")
 	}
 
-	if actual.Student != dir {
-		t.Errorf("actual.student mismatch, received %#v, want %#v", actual.Student, dir)
+	if result.Student != dir {
+		t.Errorf("actual.student mismatch, received %#v, want %#v", result.Student, dir)
 	}
 }
```

Comment:
Now that we have 4 language options, each with their own potential for unique issues, we should break down a tests more for further coverage. We could turn `TestCompile()` into `TestCppCompile()` and `TestJavaCompile()`, `TestRunCompiled()` to `TestRunCompiledCpp()` and `TestRunCompiledJava()`, etc. Not blocking, just jotting down the idea as it'll be something we'll all aim to do moving forward.

Reply:
Yes, yes I completely agree, going to implement all those tests in the next sprint. They should pretty much follow the same structure.